### PR TITLE
Turn GenericRemoteStorage into just an alias for 'Box<dyn RemoteStorage>'

### PR DESCRIPTION
**Pull request by hlinnaka**
We had a pattern like this:

    match remote_storage {
        GenericRemoteStorage::Local(storage) => {
            let source = storage.remote_object_id(&file_path)?;
            ...
            storage
                .function(&source, ...)
                .await
       },
       GenericRemoteStorage::S3(storage) => {
	    ... exact same code as for the Local case ...
       },

This removes the code duplication, by allowing you to call the functions
directly on GenericRemoteStorage.

Also change RemoveObjectId to be just a type alias for String. Now that
the callers of GenericRemoteStorage functions don't know whether they're
dealing with the LocalFs or S3 implementation, RemoveObjectId must be the
same type for both.

### CODE REVIEW

File: pageserver/src/storage_sync/index.rs [hunk 1]
```diff
@@ -278,8 +278,7 @@ pub struct IndexPart {
 }
 
 impl IndexPart {
-    pub const FILE_NAME: &'static str = "index_part";
-    pub const FILE_EXTENSION: &'static str = "json";
+    pub const FILE_NAME: &'static str = "index_part.json";
```
Comment (hlinnaka) [inline pageserver/src/storage_sync/index.rs:282]: I meant to extract this change to a separate PR but forgot. 

We always use FILE_NAME and FILE_EXTENSION together, so I didn't see the point of keeping them separate.

**Review by SomeoneToIgnore — COMMENTED** [review]
I almost like this attempt, I have to say 🙂 

I still prefer to rely on the compiler and typization more, so don't see any issues with having an explicit match and write the code a bit more.
We're able to reduce the code duplication more: 
* add a macro that has the same AST and put it into all match arms where the code is identical, but the types are not
* when all `match` starts to repeat itself in other modules, we could move this logic into `impl GenericRemoteStorage`.
We don't have to add any traits, just name the methods similarly and do the match inside

I like the flexibility and rigorousness such approach brings in terms of dealing with the different storage types though, that's what I've tried to show with the `RemoteObjectName` trait removal to demonstrate how this match could help us.

--------------

I think the main source of confusion is that S3 paths and LocalFs paths are very similar, and all code around them is similar too, so I see temptation to unify it more.

The main question then would be: do we plan to add more storages later? Might those contain non-PathBuf-like object IDs?
To me, the answer is "Yes" to both (in general, every storage has its own approach to object ids) hence the zealousness towards typization, otherwise we're stuck with `object_name` method bloated with million IFs and some disconnected parsing code at least has to take all various IDs into account.

**Comment by SomeoneToIgnore** [comment]
> we could move this logic into impl `GenericRemoteStorage`

I went on with that and extracted download and upload operations, as proposed: https://github.com/neondatabase/neon/pull/2373

I've left deletion and list_prefixes as is, with matches in the code, since they are used in the single place and quite odd operations per se, to let them be commonly used. After that PR, I personally think the old API is good enough for now: all common methods could go into `GenericRemoteStorage`, the rest are handled in-place.

File: libs/remote_storage/src/local_fs.rs [hunk 2]
```diff
@@ -85,38 +100,46 @@ impl LocalFs {
...
+    ) -> anyhow::Result<Vec<RemoteObjectId>> {
+        let x;
+        let path = match prefix {
+            Some(prefix) => {
+                x = PathBuf::from(prefix.0.clone());
+                &x
+            }
+            None => &self.storage_root,
+        };
         get_all_files(path, false).await
```
Comment (LizardWizzard) [inline libs/remote_storage/src/local_fs.rs:135]:
```suggestion
        let path = match prefix {
            Some(prefix) => Path::new(&prefix.0),
            None => &self.storage_root,
        };
        get_all_files(path, false).await
```

File: libs/remote_storage/src/lib.rs [hunk 3]
```diff
@@ -135,35 +171,23 @@ impl std::error::Error for DownloadError {}
...
-            RemoteStorageKind::LocalFs(root) => {
-                info!("Using fs root '{}' as a remote storage", root.display());
-                LocalFs::new(root.clone(), working_directory).map(GenericRemoteStorage::Local)
-            }
-            RemoteStorageKind::AwsS3(s3_config) => {
-                info!("Using s3 bucket '{}' in region '{}' as a remote storage, prefix in bucket: '{:?}', bucket endpoint: '{:?}'",
-                    s3_config.bucket_name, s3_config.bucket_region, s3_config.prefix_in_bucket, s3_config.endpoint);
-                S3Bucket::new(s3_config, working_directory).map(GenericRemoteStorage::S3)
-            }
+pub type GenericRemoteStorage = Box<dyn RemoteStorage>;
```
Comment (LizardWizzard) [inline libs/remote_storage/src/lib.rs:174]: Frequently we use `Arc<GenericRemoteStorage>` so this ends up being `Arc<Box<dyn RemoteStorage>>`. Arc -> Box becomes a redundant indirection, not that it will hurt, but just to keep things clean

I think this way it worth making `GenericRemoteStorage` actual newtype around `Arc<dyn RemoteStorage>` and implement `Clone` so it calls to `Arc::clone`. Wdyt?

**Review by LizardWizzard — APPROVED** [review]
Looks nice to me. Left some minor comments

File: libs/remote_storage/src/lib.rs (hunk 3 above)
Comment (hlinnaka) [reply libs/remote_storage/src/lib.rs:174]: Works for me.

You want to take over this PR and rebase / clean it up to make it committable?

File: libs/remote_storage/src/lib.rs [hunk 4]
```diff
@@ -42,57 +39,96 @@ pub const DEFAULT_REMOTE_STORAGE_MAX_SYNC_ERRORS: u32 = 10;
...
+/// and remote keys.
+///
+impl RemoteObjectId {
+    // Needed to retrieve last component for RemoteObjectId.
+    // In other words a file name
+    /// Turn a/b/c or a/b/c/ into c
+    pub fn object_name(&self) -> Option<&str> {
+        // corner case, char::to_string is not const, thats why this is more verbose than it needs to be
+        // see https://github.com/rust-lang/rust/issues/88674
+        if self.0.len() == 1 && self.0.chars().next().unwrap() == S3_PREFIX_SEPARATOR {
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply libs/remote_storage/src/lib.rs:59]: Hmm. I think our disagreements here stem from different views on what RemoteObjectId represents. In my mind, RemoteObjectId is a "name" or "key" that can be used to refer to a remote object. It's not just an internal matter of the RemoteStorage implementation, it's something that the users of RemoteStorage need to construct and deal with. For example, to upload a file, you construct a RemoteObjectId, and then you upload the file with that ID. And later you can retrieve it back by using the same ID.

I think you're thinking of it as more of a physical locator in the remote storage. For the LocalFs implementation, it's the absolute path in the local filesystem to the file, and for S3 it's the prefix+key in S3. Am I right?

To me, it would feel more natural if RemoteObjectId only contained the caller-supplied part. For example, if you upload a file as `tenant1/timeline2/index.part` with the LocalFs implementation, RemoteObjectId should wrap just `tenant1/timeline2/index.part`, not `/data/storage/tenant1/timeline2/index.part`. The implementation will join the storage root and the RemoteObjectId when it opens the file for reading/writing. Similarly for S3, the bucket or prefix should not be part of RemoteObjectId.

That way, a RemoteObjectId would not be tied to a particular RemoteStorage implementation, or even to a particular RemoteStorage object. For example, if you wanted to copy a file from one RemoteStorage to another, you could do something like this:
```
let remote_object_id = RemoteObjectId("foo/bar");
remote_storage_src.download(remote_object_id);
remote_storage_dst.upload(remote_object_id);
```

Thoughts?

(Yeah, the code on RemoteObjectId should not involve S3_PREFIX_SEPARATOR.)

[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply libs/remote_storage/src/lib.rs:59]: Yes, I think you've got the idea right and I like this new proposal too, as something I've tried to describe [here](https://github.com/neondatabase/neon/pull/2387#issuecomment-1236876167).

My idea was to still keep the separation though, since the object pointers are different for every storage, but to keep some sort of a generic key, that could be converted into the "specific" key later by the repo itself.

[the remark this answers is not in the document]
Comment (hlinnaka) [reply libs/remote_storage/src/lib.rs:59]: > My idea was to still keep the separation though, since the object pointers are different for every storage, but to keep some sort of a generic key, that could be converted into the "specific" key later by the repo itself.

I'm cool with that

File: libs/remote_storage/src/lib.rs (hunk 3 above)
Comment (LizardWizzard) [reply libs/remote_storage/src/lib.rs:174]: Sure

**Comment by LizardWizzard** [comment]
@SomeoneToIgnore I've rebased this PR on top of your changes, what do you think about it in the current state?

**Review by SomeoneToIgnore — APPROVED** [review]
It's quite an interesting approach with `Deref`, actually (despite some talks of it as an [antipattern](https://github.com/rust-unofficial/patterns/blob/main/anti_patterns/deref.md)).

Disregarding the concerns in the link, I would still argue against the general approach: we already start to bend the `RemoteStorage` __public__ interface with `as_local()` method.
With the deref to `&dyn RemoteStorage` approach chosen, we have no way but to add more odd methods to deal with storage differences, when the need appears (I hope it won't, not in midterm at least).

------

My proposal would be still to explicitly differentiate the storage types with enum, and write some adapters per storage type, writing some macros to deduplicate the ast if that looks so bothersome.

`RemoteObjectId` and `GenericRemoteStorage` could act as reexported public interfaces, with all the enum matching glue code hidden in a separate module with them.
`RemoteStorage` would be a __private__ trait, that helps us to deal with common storage operations we need them to share in the glue code.

`fn remote_object_id` and `fn local_path` shoud not deal with `&Path` (it's odd to have it there even before this PR, come to think of it now) but rather with `&RemoteObjectId` thus enabling us to write that glue code almost identically now.

We get code deduplication, one public object with sane helper methods and nothing else leaked in the impl.
We can also hide the enum with storages too, hiding entire `RemoteStorage` and converters from the public API.

------

Yet, this code does set a good standard for zero weird matches in the API user code, brings in `RemoteObjectId` (in an odd way now, as I've tried to explain), but still.
Finally, I'm thinking that my argumentation drains too much of attention from other, more pressing, issues.
I would love to come up with another counter-PR, implementing my proposal above, but cannot do it in the very near future, since started to step on tenants' issues: https://github.com/neondatabase/neon/pull/2406 , https://github.com/neondatabase/neon/pull/2395

For now, I guess, this is some sort of a compromise so why not try it.

**Comment by LizardWizzard** [comment]
> It's quite an interesting approach with Deref, actually (despite some talks of it as an [antipattern](https://github.com/rust-unofficial/patterns/blob/main/anti_patterns/deref.md)).

From the link:

> The Deref trait is designed for the implementation of custom pointer types. The intention is that it will take a pointer-to-T to a T, not convert between different types. It is a shame that this isn't (probably cannot be) enforced by the trait definition.

Here it acts like an alias to Arc<dyn RemoteStorage> but since you cant have an impl on aliases it is a newtype. So in some sense it is a pointer type. An alternative is to implement RemoteStorage for GenericRemoteStorage as Heikki suggested in other PR.

> brings in RemoteObjectId (in an odd way now, as I've tried to explain), but still

I agree with your arguments, I like the above Heikki's [proposal](https://github.com/neondatabase/neon/pull/2368#discussion_r963560686) regarding RemoteStorageId. Would be good to explore that. To me it is related but a separate question from dyn vs enum. I do not strictly opposed to enum approach, I think with this remote object id streamlining both dyn or enum approach will look better.

If there are no objections I would rather merge this and continue exploring for possible improvements

**Comment by SomeoneToIgnore** [comment]
Not objecting against anything related at this point.
